frontend/tests/unit/test_source_generator.py

```diff
@@ -0,0 +1,149 @@
+import re
+from filecmp import dircmp
+import sh
+from frontend_benchmark.user_callgraph import Callgraph
+from frontend_benchmark.source_generator import SourceGenerator
+
+
+@pytest.fixture
+def rootdir():
+    return os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                        os.path.pardir)
+
+
+@pytest.fixture
+def resources(rootdir):
+    return os.path.join(rootdir, 'resources')
+
+
+def test_branch_indirect_call(resources, tmpdir):
+    test_file = os.path.join(resources, 'branch_indirect_call.pbtext')
+    cfg = Callgraph.from_proto(test_file)
+    source_gen = SourceGenerator(tmpdir, cfg)
+    source_gen.write_files()
+    output = compile_and_return_output(
+        tmpdir, 'function_2', switch_processor(r'callq\s+\*', r'\s+blr\s+x\d+'))
+    assert output, 'Missing indirect function call'
+
+
+def test_branch_direct_call(resources, tmpdir):
+    test_file = os.path.join(resources, 'branch_direct_call.pbtext')
+    cfg = Callgraph.from_proto(test_file)
+    source_gen = SourceGenerator(tmpdir, cfg)
+    source_gen.write_files()
+    output = compile_and_return_output(
+        tmpdir, 'function_2', switch_processor(r'callq\s+\d*', r'\s+bl\s+\d*'))
+    assert output, 'Missing direct function call'
+
+
+def test_branch_indirect(resources, tmpdir):
+    test_file = os.path.join(resources, 'branch_indirect.pbtext')
+    cfg = Callgraph.from_proto(test_file)
+    source_gen = SourceGenerator(tmpdir, cfg)
+    source_gen.write_files()
+    output = compile_and_return_output(
+        tmpdir, 'function_2', switch_processor(r'jmp\w*\s+\*', r'\s+br\s+x\d+'))
+    assert output, 'Missing indirect branch'
+
+
+def test_branch_conditional_direct(resources, tmpdir):
+    test_file = os.path.join(resources, 'branch_conditional_direct.pbtext')
+    cfg = Callgraph.from_proto(test_file)
+    source_gen = SourceGenerator(tmpdir, cfg)
+    source_gen.write_files()
+    output = compile_and_return_output(
+        tmpdir, 'function_2', switch_processor(r'\s+je\s+.*',
+                                               r'\s+b\.\w+\s+\d+'))
+    assert output, 'Missing indirect branch'
+
+
+def test_branch_implicit_fallthrough(resources, tmpdir):
+    test_file = os.path.join(resources, 'branch_implicit_fallthrough.pbtext')
+    cfg = Callgraph.from_proto(test_file)
+    source_gen = SourceGenerator(tmpdir, cfg)
+    source_gen.write_files()
+    output = compile_and_return_output(
+        tmpdir, 'function_2', switch_processor(r'\s+je\s+.*',
+                                               r'\s+b\.\w+\s+\d+'))
+    assert output, 'Implicit fallthrough branch error'
+
+
+def switch_processor(value1, value2):
+    processor = platform.processor()
+    if processor == 'x86_64':
+        return value1
+    elif processor == 'aarch64':
+        return value2
+    raise RuntimeError(f'Unrecognized processor: {processor}')
+
+
+def compile_and_return_output(directory, symbol, asm):
+    """ WARNING. THIS ASSUMES RUNNING ON x86 """
```

Comment:
Which part of this assumes x86? Isn't the gcc command also named "gcc" on ARM (and similarly for objdump)?

Reply:
Carryover from old code. Deleted.